Refactor lowering

csarofeen opened the pull request:
Fully functional. Still some todo on cleaning. Only place left for cleanup is compute at maps. I want to narrow down precisely what's needed there and remove some unused items.

csarofeen committed:
Move expression sorting out of lower loops.

csarofeen committed:
Draft new sorting mechanism based on segmentation pass.

csarofeen committed:
Pipe through compute at map, start to move to it.

csarofeen committed:
Extend compute at map to find max size concretized domains, and to kernel IR iter domains.

csarofeen committed:
Test loop lowering with compute at map, instead of computeAt structures.

csarofeen committed:
Move sync threads logic out of lower loops into its own pass.

csarofeen committed:
Missed cleanup of a member.

csarofeen committed:
Rewrite raw sync logic, as it seems ther were some cases missing, logic also seems more clear now, also add debug print for vector of kir::Expr.

csarofeen committed:
Move allocation into its own pass, allocate closer to where required.

csarofeen committed:
Close to functional.

csarofeen committed:
Minor cleanup. Test fix. Index Fix.

csarofeen committed:
First example working SimpleBCast1

csarofeen committed:
Just 30 failing tests left.

csarofeen committed:
Split out loop and index map, further debugging.

csarofeen committed:
Merge branch '20_12_3_devel' of https://www.github.com/csarofeen/pytorch into indexRefactor

csarofeen committed:
Clang format.

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_iostream.cpp:
```diff
@@ -68,7 +68,8 @@ void IrPrinter::handle(const TensorView* tv) {
     if (tv->getComputeAtView() != nullptr) {
       os_ << " compute_at( ";
       os_ << "T" << tv->getComputeAtView()->name();
```
Can be reverted once we change computeAt structure, but for now should leave in for debugging.

csarofeen commented on torch/csrc/jit/codegen/cuda/kernel_ir_printer.cpp:
```diff
@@ -1,6 +1,7 @@
 #include <torch/csrc/jit/codegen/cuda/kernel_ir_printer.h>
 
 #include <torch/csrc/jit/codegen/cuda/instrumentation.h>
+#include <torch/csrc/jit/codegen/cuda/ir_iostream.h>
```
changes in this file should be removed, or split into another PR, still needed for debugging.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_allocation.cpp:
```diff
@@ -113,10 +112,12 @@ class AllocationInserter : public kir::MutableIrVisitor {
+                  fuser_tv->axis(axis_i)))
               ->as<kir::IterDomain>();
-      init_dims.push_back(ca_id);
+      init_dims.push_back(concrete_id);
```
@naoyam is this right? Should we allocate/init based on a non-local iteration domain size? What happens when compute at axis is non-broadcast but local id is?

naoyam replied:
It should be safe to ignore broadcast IDs. I suspect that isn't simply taken into consideration in our current implementation.

This may be related: #530

naoyam replied:
Hmm, broadcast *is* ignored in `createAllocExpr`. Maybe there is some bug around here.

naoyam replied:
It seems that we do create an initialization expression with a broadcast loop but don't generate it.

https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/codegen.cpp#L593

So, it's no harm, but seems to make more sense to skip such domains at this point.

naoyam replied:
I'll create a small PR to ignore broadcast domains. 

Aside from the case of local-broadcast axes, is there any other concern here?

naoyam replied:
Please see #601

csarofeen replied:
Thank you.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_allocation.cpp:
```diff
@@ -195,7 +200,10 @@ class AllocationInserter : public kir::MutableIrVisitor {
       }
-      alloc_dims.push_back(ca_id->rawExtent());
+      // TODO: Is this correct? Should allocations be based off of concrete
+      // sizes?
```
@naoyam same thing here.

csarofeen replied:
Comments posted in reverse, see below, sorry.

csarofeen replied:
I think this should probably be indexMap concrete, not loop map

csarofeen replied:
TODO: change, run memcheck

csarofeen replied:
Remove TODO, concrete size in index map cannot be larger than local size.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_utils.cpp:
```diff
@@ -168,9 +168,12 @@ ParallelTypeBitmap getParallelBroadcastDomains(
 namespace loop_utils {
 
+// TODO: Clean this up, Naoya added a mechanism we should be able to reuse.
 std::pair<kir::ForLoop*, int64_t> getAllocPoint(
```
@naoyam could you please double check the logic in here, worried I may be messing up the alloc point like I did with the lower alloc pass.

naoyam replied:
This looks good to me.

csarofeen commented on torch/csrc/jit/codegen/cuda/predicate_compute.cpp:
```diff
@@ -113,7 +113,6 @@ kir::Bool* PredicateCompute::getInlinePredicate(
     const ComputeAtRootDomainMap& ca_root_map,
     bool ignore_block_grid_reductions) {
   FUSER_PERF_SCOPE("getInlinePredicate");
-
```
Minor changes should be reverted.

csarofeen committed:
Use maybeRfactor when indexing consumers.

csarofeen committed:
Remove ca root map during lowering. It shouldn't be needed. Leave it currently as a stored member of GpuLower so its easier to find its uses.

csarofeen committed:
Don't reuse iteration domains in arith, make sure they're cloned.

csarofeen committed:
Minor rename.

csarofeen committed:
Only 3 failing tests left.

csarofeen committed:
Clang format.

csarofeen committed:
All tests working, still need to reimplement swizzle.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -353,9 +354,9 @@ TEST(NVFuserTest, FusionExprEvalPostLower_CUDA) {
-  tv3->axis(1)->parallelize(ParallelType::Unroll);
+  // tv3->axis(0)->parallelize(ParallelType::BIDx);
+  // tv2->axis(1)->parallelize(ParallelType::Unroll);
+  // tv3->axis(1)->parallelize(ParallelType::Unroll);
```
Revert

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -3956,6 +3957,11 @@ TEST(NVFuserTest, FusionSimpleBCast1_CUDA) {
+  for (auto tv : tvs) {
+    // tv->setMemoryType(MemoryType::Global);
+  }
+
```
Revert

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -4599,6 +4605,9 @@ TEST(NVFuserTest, FusionAdvancedIndexing7_CUDA) {
   tv4->axis(0)->parallelize(ParallelType::TIDx);
 
+  fusion.printMath();
+  fusion.printKernel();
```
Revert

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -4618,6 +4627,7 @@ TEST(NVFuserTest, FusionAdvancedIndexing7_CUDA) {
       &fusion, cg_outputs, {at_t0, at_t1}, {aten_output}, __LINE__, __FILE__);
 }
 
+// TODO: Enable test
```
TODO

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -4639,30 +4649,84 @@ TEST(NVFuserTest, FusionAdvancedIndexing8_CUDA) {
+  //     __FILE__);
+}
+
+// TODO: Enable test
```
TODO

csarofeen replied:
Requires changes to root mapping. CC @naoyam . Probably should do this in a follow up.

csarofeen commented on torch/csrc/jit/codegen/cuda/executor.cpp:
```diff
@@ -138,6 +138,7 @@ void FusionExecutor::compileFusion(Fusion* fusion, CompileOptions options) {
   }
 
   const auto kernel_code = codegen::generateCudaKernel(kernel, kernelName());
+  // std::cout<<kernel_code<<std::endl;
```
Revert

csarofeen commented on torch/csrc/jit/codegen/cuda/index_compute.cpp:
```diff
@@ -429,8 +452,9 @@ IndexCompute IndexCompute::updateIndexCompute(
 
   const auto gpu_lower = GpuLower::current();
```
TODO: remove new_index_entries in this function

csarofeen commented on torch/csrc/jit/codegen/cuda/index_compute.cpp:
```diff
@@ -659,16 +699,16 @@ void IndexSwizzle::run() {
       auto idx_to_swizzle_j = indexMap().at(id_to_swizzle_j_kir);
 
-      kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+      kir::IrBuilder ir_builder(gpu_lower->kernel());
```
Move up with gpu_lower definition for consistency

csarofeen commented on torch/csrc/jit/codegen/cuda/index_compute.cpp:
```diff
@@ -695,410 +735,61 @@ void IndexSwizzle::handle(Expr* e) {
-  // grab all tensor views from producer_tv <- computeAtRoot
-  auto tv_stack = getComputeAtTVStackFrom(consumer_tv);
-  tv_stack.push_back(producer_tv);
+  // Cannot be done after domain_guard
```
TODO: Check this, domain guard shouldn't impact root mapping.

csarofeen commented on torch/csrc/jit/codegen/cuda/index_compute.cpp:
```diff
@@ -1209,28 +902,112 @@ std::unordered_map<kir::ForLoop*, kir::Val*> indexMapFromTV(
-      std::vector<bool>(producer_tv->getRootDomain().size(), false),
-      ca_root_map,
-      true,
+  // Cannot be done after domain_guard
```
TODO: Also check.

csarofeen commented on torch/csrc/jit/codegen/cuda/index_compute.cpp:
```diff
@@ -1384,33 +1179,139 @@ kir::TensorIndex* Index::getGlobalConsumerIndex(
+  // they can be consumed) and since global consumers won't have zero merge in
+  // propagation this can only happen here. We will go through and zero out any
+  // for loops that are a reduction axis for the consumer, first find them.
+  {
```
TODO: compare with producer for correctness

csarofeen commented on torch/csrc/jit/codegen/cuda/index_compute.cpp:
```diff
@@ -1384,33 +1179,139 @@ kir::TensorIndex* Index::getGlobalConsumerIndex(
+  // Map reference tensor to consumer
+  std::unordered_map<IterDomain*, IterDomain*> root_ref_to_consumer;
+  for (auto c_root : consumer_tv->getMaybeRFactorDomain()) {
+    auto concrete_id = gpu_lower->caIndexMap().getConcreteMappedID(c_root);
```
TODO: Is this okay for consumer, though not for producer?

csarofeen commented on torch/csrc/jit/codegen/cuda/index_compute.cpp:
```diff
@@ -1384,33 +1179,139 @@ kir::TensorIndex* Index::getGlobalConsumerIndex(
+      {},
+      consumer_tv->domain()->contiguity());
+
+  // TODO: Re-enable swizzle
```
TODO: Swizzle.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower2device.cpp:
```diff
@@ -126,21 +148,21 @@ void GpuLower::lower() {
   // corresponding loop
-  const auto lowered_exprs =
-      LoopNestGenerator::loweredExprs(fusion_, reordered_exprs);
-
```
revert const

csarofeen commented on torch/csrc/jit/codegen/cuda/lower2device.h:
```diff
@@ -36,6 +38,22 @@ class TORCH_CUDA_API GpuLower {
   //! (or nullptr if no lowering is in progress)
   static GpuLower* current();
 
+  const ComputeAtRootDomainMap& caRootMap() const {
```
Do we still use this anywhere? Would be good if it was non ca version.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_allocation.cpp:
```diff
@@ -113,10 +112,12 @@ class AllocationInserter : public kir::MutableIrVisitor {
               ->as<kir::IterDomain>();
-      init_dims.push_back(ca_id);
+      init_dims.push_back(concrete_id);
     }
```
TODO: make sure this comes through https://github.com/csarofeen/pytorch/pull/601

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp:
```diff
@@ -0,0 +1,658 @@
+  return parallel_type_it->second;
+}
+
+// std::unordered_map<IterDomain*, IterDomain*> ComputeAtMap::mapFromTo(
```
TODO: Cleanup

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_loops.cpp:
```diff
@@ -221,6 +224,185 @@ void LoopNestGenerator::generate(const std::vector<Expr*>& exprs) {
     handle(expr);
   }
 }
+// ======================================================
```
TODO: Clean up.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1,4 +1,4 @@
-#if defined(USE_CUDA)
+// #if defined(USE_CUDA)
```
Revert

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -11424,7 +11491,64 @@ TEST(NVFuserTest, FusionTransposeWithSwizzle1DThreadBlock_CUDA) {
 } // namespace torch
 
-#endif // #if defined(USE_CUDA)
+// #endif // #if defined(USE_CUDA)
```
Revert

csarofeen commented on tools/build_variables.bzl:
```diff
@@ -371,6 +371,7 @@ libtorch_cuda_sources = [
     "torch/csrc/jit/codegen/cuda/fusion.cpp",
     "torch/csrc/jit/codegen/cuda/graph_fuser.cpp",
     "torch/csrc/jit/codegen/cuda/index_compute.cpp",
+    "torch/csrc/jit/codegen/cuda/test_index_compute.cpp",
```
TODO: Rename

csarofeen commented on tools/build_variables.bzl:
```diff
@@ -405,6 +407,7 @@ libtorch_cuda_sources = [
     "torch/csrc/jit/codegen/cuda/register_interface.cpp",
     "torch/csrc/jit/codegen/cuda/root_domain_map.cpp",
     "torch/csrc/jit/codegen/cuda/scheduler.cpp",
+    "torch/csrc/jit/codegen/cuda/segmenter_helper.cpp",
```
TODO: Rename

csarofeen committed:
Minor cleanup.

csarofeen committed:
Hook swizzle back up.

csarofeen committed:
Test cleanup.

csarofeen committed:
Always use root domain mapping, don't try to implicitly line them up!

csarofeen committed:
Clean up index compute.

csarofeen committed:
Minor cleanup.

csarofeen committed:
Clean up reference replay code.

csarofeen committed:
Move segmentation logic directly into expression sorting file.

csarofeen commented on torch/csrc/jit/codegen/cuda/kernel_ir_printer.h:
```diff
@@ -21,12 +21,15 @@ namespace kir {
 //! This class is intended for debug printing, so it attempts
```
TODO: Revert changes, or split out to a separate PR.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower2device.cpp:
```diff
@@ -109,14 +109,30 @@ void GpuLower::lower() {
-  ca_root_map.build();
+  // In the future we may directly use this map, but for now it will propagate
+  // and validate (to some extent) the parallelization strategy.
+  // TODO: Does this have to go before the next mapping?
```
Fix comment, no longer TODO, has to be done first since right now we're setting parallel strategy on fusion iter domains and need to do it before lowering.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_loops.cpp:
```diff
@@ -17,17 +18,22 @@ namespace jit {
 
 // TODO(kir): revisit and try to simplify this
-kir::ForLoop* openForHelper(kir::ForLoop* scope, IterDomain* id) {
+kir::ForLoop* openForHelper2(kir::ForLoop* scope, IterDomain* id) {
```
TODO: Fix name back.

csarofeen committed:
Cleanup expression sorting, remove generalization of segmentation.

csarofeen committed:
Minor cleanup expr sorting.

csarofeen committed:
Cleanup of compute at maps.

csarofeen committed:
Remove some old cout statements.

csarofeen committed:
Add some minor modifications to kernel printing.

csarofeen committed:
Revert printing changes as they're separated out into another PR.

csarofeen committed:
Minor cleanup.

csarofeen committed:
Minor cleanup.

csarofeen committed:
Clang tidy: pass 1

csarofeen committed:
Clang tidy: pass 2

csarofeen committed:
Clang tidy: pass 3

Commit:
minor cleanup

Commit:
Merge branch 'minor_printing_changes' into loweringRefactor

Commit:
Check size before front

Commit:
Merge branch 'fix-insert-sync' into loweringRefactor

Commit:
cleanup

Commit:
cleanup

Commit:
cleanup

naoyam commented on torch/csrc/jit/codegen/cuda/lower_utils.cpp:
```diff
@@ -168,9 +168,12 @@ ParallelTypeBitmap getParallelBroadcastDomains(
 
 namespace loop_utils {
 
+// TODO: Clean this up, Naoya added a mechanism we should be able to reuse.
```
Which mechanism are you referring to?

csarofeen replied:
I thought we may want to reuse `lower_allocation.cpp::findAllocationPosition`

naoyam replied:
I'll try to refactor them.

naoyam commented on torch/csrc/jit/codegen/cuda/lower_loops.cpp:
```diff
@@ -88,137 +103,83 @@ void LoopNestGenerator::handle(const Expr* expr) {
-  // Lets get a copy of the loop structure, and figure out which loops we need
-  // to open.
-  auto loops_to_open = loop_structure;
+  // Save position of out_id_it as we will append to loop structure
```
Shouldn't this be placed before the above for loop (i.e., before line 148)?

csarofeen replied:
Yes, it should be.

csarofeen replied:
Moved.

naoyam commented on torch/csrc/jit/codegen/cuda/lower_loops.cpp:
```diff
@@ -88,137 +103,83 @@ void LoopNestGenerator::handle(const Expr* expr) {
-    }
+    // Safe to use loop map since this is outside the compute at point
+    auto concrete_id =
+        gpu_lower->caParallelMap().getConcreteMappedID(out_tv->axis(out_i));
```
Shouldn't this be `caLoopMap()` instead of `caParallelMap()`?

csarofeen replied:
I actually think it should be index or parallel. Within the compute at they should have the maximum extent that would be needed.

csarofeen replied:
No, you're right, it should be loop because of the misaligned bcast behavior.

naoyam replied:
Hmm, I tried `caLoopMap()` and got a validation error with `FusionPersistentNormLocalShared`. Will look into it.

csarofeen replied:
Parallel map should be fine for now. Although in the future we'll want to move to generating a reference either before or during the loop construction and including the reference iter domains in the iter domain maps. The issue that could happen from here is if one of the TensorViews doesn't resolve all the broadcasts that are resolved in a loop nest. This doesn't happen in our current testing but could be done.

naoyam commented on torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp:
```diff
@@ -0,0 +1,578 @@
+          }
+          // Map the id's together
+          map_ids(p_id, c_id);
+          p2c_map[p_id] = c_id;
```
c2p_map is not one-to-one when the mapping mode is LOOP, i.e., when bcast mismatch is allowed. Doesn't it cause p2c_map non-deterministic? There can be multiple c_id for a single p_id. The final mapping for the p_id would depend on the internal ordering of c2p_map.

csarofeen replied:
There are multiple p_id's it could map to, but the mapping will simply forward from best effort replay. So the most bcast merged axis in producer should be mapped to the corresponding c_id.
I'm not 100%, but I'm pretty sure it should be deterministic.

naoyam replied:
Apparently, there are multiple different c_ids for a single p_id. I inserted this:

```
diff --git a/torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp b/torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp
index 6001e02dc5..c320ee9781 100644
--- a/torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp
+++ b/torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp
@@ -287,6 +287,7 @@ void ComputeAtMap::build() {
           }
           // Map the id's together
           mapIds(p_id, c_id);
+          TORCH_INTERNAL_ASSERT(p2c_map.find(p_id) == p2c_map.end() || p2c_map.at(p_id) == c_id);
           p2c_map[p_id] = c_id;
         }
```

And I got assertion failures with it. So, the mapping of the p_id would just depend on the ordering among the multiple c_ids, which would be non-deterministic.

csarofeen replied:
Fixed this with https://github.com/csarofeen/pytorch/pull/597/commits/6f279a2ecb5a135b4ef9d99fbfacb75af60c8673

Commit:
Cleanup

naoyam commented on torch/csrc/jit/codegen/cuda/lower_expr_sort.h:
```diff
@@ -7,9 +7,9 @@ namespace jit {
 namespace cuda {
 
-std::vector<Expr*> reorderExprsForComputeAt(const std::vector<Expr*>& exprs);
+TORCH_CUDA_API std::vector<Expr*> reorderExprsTest();
```
Is there any reason to name this function `-Test`?

csarofeen replied:
No of course not, just forgot to rename it.

csarofeen replied:
Reverted name.

Commit:
Add underscore suffix

naoyam commented on torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp:
```diff
@@ -1,372 +1,904 @@
+  auto it1 = domain1.begin();
+  auto it2 = domain2.begin();
+
+  while (it1 != domain1.end() && it2 != domain2.end()) {
```
Shouldn't this use `||` instead of `&&`? With `&&`, the first two if clauses are never used.

naoyam replied:
Can we just do something like:

```
auto resulting_ca_axes = domain1;
for (auto id2 : domain2) {
  if (std::none_of(domain1.begin(), domain1.end(), [&](auto id1) {
         return caIndexMap().areMapped(id1, id2);
      })) {
      resulting_ca_axes.push_back(id2);
  }
}
```

csarofeen replied:
We need all iteration domains from both to propagate, and we need to maintain relative order while trying to reduce duplicates when we can. I added some more comments.

csarofeen replied:
You were right about the `||` instead of `&&`, changed that as well.

Commit:
refactoring

Commit:
cleanup

csarofeen committed:
Fixes from comments.

Commit:
cleanup

Commit:
Merge remote-tracking branch 'origin/loweringRefactor' into loweringRefactor

Commit:
clang-format

Commit:
Use the loop map instead of the index map as the sorting is for
loop-nest generation

csarofeen committed:
Make sure produce at is correct and deterministic.

csarofeen committed:
Merge branch 'loweringRefactor' of https://www.github.com/csarofeen/pytorch into indexRefactor

Commit:
cleanup

Commit:
Add an assertion that I believe should never be reachable.

Will revisit the logic when the assertion fails.

Commit:
cleanup

Commit:
Add a comment about the assertion

Commit:
clang-tidy

Commit:
cleanup

naoyam commented on torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp:
```diff
@@ -1,372 +1,963 @@
+
+  // TODO: Used?
+  std::vector<Val*> input_vals_;
+  std::vector<Val*> output_vals_;
```
Are these used? They are set, but they do not seem to be used. Can they be removed?

csarofeen replied:
If they're not used, then yes, please remove.

naoyam replied:
Removed.

Commit:
Remove unused vals

Commit:
Forward missing bcast in the parallel map as well

Parallelization is defined at each loop, so it should behave similarly
with the loop map.

Commit:
Use parallel map when creating alloc exprs.

There should be no effective impact with this change, but I'd say this
would be simpler to understand.

Commit:
cleanup

Commit:
Fix adding remaining IDs

Commit:
Error out when multiple loops are mapped to the same concrete ID

Commit:
Remove checks of broadcast axes for replayed domains when forwarding
mismatched merge

Commit:
Replaying the reference tensor without forwarding mismatched bcast

naoyam commented on torch/csrc/jit/codegen/cuda/index_compute.cpp:
```diff
@@ -444,6 +465,10 @@ IndexCompute IndexCompute::updateIndexCompute(
       updated_index_map[new_id] = index_map_.at(prev_id);
     }
 
+    if (!prev_id->isBroadcast() && new_id->isBroadcast()) {
```
Can you give some comments here? `updated_extent_map` is set at three different paths with `extent_map_.at(prev_id)` or `getExtent(prev_id)`. Would be nice why each path would be taken and why it would use `extent_map_` or `getExtent`.

csarofeen replied:
I think this should actually be an `if else` in the `if` below. This is primarily just not setting an entry in the extent map unless it's needed. That way we know if extent has diverged from what's in the `IterDomain`. `getExtent` is just a convenience wrapper that will return a map entry if there is one, otherwise will return the extent of the `IterDomain` provided.

csarofeen replied:
The one I actually don't get below is the is reduction section, uncertain if or why that'd be needed. Another thing to keep in mind, is I think we only update the index compute once per indexing now, instead of through out a large tv stack so we may be able to simplify this code somewhat.

naoyam replied:
Thanks. I think this part of the change is also related. The propagation through the `isReduction()` logic may not kick in now, and that seems to explain why an axis can be missing in `extent_map` .

naoyam replied:
I'll revisit this later again.

Commit:
Adds a test exercisizing the preferred path in indexing

Commit:
Add an assertion

naoyam commented:
I'm mostly done reviewing this PR. There are mainly three things to do, all of which can be done as separate PRs.

- Enable the AdvancedLowering3 test. Root mapping seems to have an issue. https://github.com/csarofeen/pytorch/pull/597#discussion_r560349607
- Refactoring of allocation position finders. https://github.com/csarofeen/pytorch/pull/597#discussion_r562038955
- IndexCompute::updateIndexCompute. https://github.com/csarofeen/pytorch/pull/597#discussion_r566519607

We may also revisit this: https://github.com/csarofeen/pytorch/pull/597#discussion_r563782875.

Commit:
Merge branch '20_12_3_devel' into loweringRefactor

Commit:
cleanup

Commit:
clang-format

Commit:
Merge fix
